Essentials/src/main/java/com/earth2me/essentials/commands/Commandtpaccept.java

```diff
@@ -18,60 +24,105 @@ public Commandtpaccept() {
+        } else {
+            user.sendMessage(tl("requestAccepted"));
+            handleTeleport(user, user.getNextTpaRequest(true, false, false), commandLabel);
+        }
+        throw new NoChargeException();
+    }
+
+    @Override
+    protected List<String> getTabCompleteOptions(Server server, User user, String commandLabel, String[] args) {
+        if (args.length == 1) {
+            final List<String> options = new ArrayList<>(user.getPendingTpaKeys());
+            options.add("*");
+            return options;
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    private void handleTeleport(final User user, final IUser.TpaRequest request, String commandLabel) throws Exception {
+        if (request == null) {
             throw new Exception(tl("noPendingRequest"));
         }
+        final User requester = ess.getUser(request.getRequesterUuid());
 
-        if (user.isTpRequestHere() && ((!requester.isAuthorized("essentials.tpahere") && !requester.isAuthorized("essentials.tpaall")) || (user.getWorld() != requester.getWorld() && ess.getSettings().isWorldTeleportPermissions() && !user.isAuthorized("essentials.worlds." + user.getWorld().getName())))) {
+        if (!requester.getBase().isOnline()) {
+            user.removeTpaRequest(request.getName());
             throw new Exception(tl("noPendingRequest"));
         }
 
-        if (!user.isTpRequestHere() && (!requester.isAuthorized("essentials.tpa") || (user.getWorld() != requester.getWorld() && ess.getSettings().isWorldTeleportPermissions() && !user.isAuthorized("essentials.worlds." + requester.getWorld().getName())))) {
+        if (request.isHere() && ((!requester.isAuthorized("essentials.tpahere") && !requester.isAuthorized("essentials.tpaall")) || (user.getWorld() != requester.getWorld() && ess.getSettings().isWorldTeleportPermissions() && !user.isAuthorized("essentials.worlds." + user.getWorld().getName())))) {
             throw new Exception(tl("noPendingRequest"));
         }
 
-        if (args.length > 0 && !requester.getName().contains(args[0])) {
+        if (!request.isHere() && (!requester.isAuthorized("essentials.tpa") || (user.getWorld() != requester.getWorld() && ess.getSettings().isWorldTeleportPermissions() && !user.isAuthorized("essentials.worlds." + requester.getWorld().getName())))) {
             throw new Exception(tl("noPendingRequest"));
         }
 
-        if (!user.hasOutstandingTeleportRequest()) {
-            user.requestTeleport(null, false);
-            throw new Exception(tl("requestTimedOut"));
+        final TeleportRequestAcceptEvent event = new TeleportRequestAcceptEvent(user, requester, request);
+        Bukkit.getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            if (ess.getSettings().isDebug()) {
+                logger.info("TPA accept canceled by api for " + user.getName());
```

Comment:
```suggestion
                logger.info("TPA accept cancelled by API for " + user.getName());
```
Probably want to include the name of the other player in the request that got cancelled here